components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java

```diff
@@ -279,12 +294,24 @@ public void storeAccessToken(String accessToken, String consumerKey,
                     prepStmt.execute();
                 }
             }
-
+            if(retryAttempt > 0) {
+                log.info("Successfully recovered 'CON_APP_KEY' constraint violation with the attempt : " +
+                        retryAttempt);
+            }
         } catch (SQLIntegrityConstraintViolationException e) {
-            String errorMsg = "Access Token for consumer key : " + consumerKey + ", user : " +
-                              accessTokenDO.getAuthzUser() + " and scope : " +
-                              OAuth2Util.buildScopeString(accessTokenDO.getScope()) + "already exists";
-            throw new IdentityOAuth2Exception(errorMsg, e);
+
+            if (retryAttempt >= tokenPersistRetryCount) {
+                log.error("'CON_APP_KEY' constrain violation retry count exceeds above the maximum count - " +
+                        tokenPersistRetryCount);
+                String errorMsg = "Access Token for consumer key : " + consumerKey + ", user : " +
+                        accessTokenDO.getAuthzUser() + " and scope : " +
+                        OAuth2Util.buildScopeString(accessTokenDO.getScope()) + "already exists";
+                throw new IdentityOAuth2Exception(errorMsg, e);
+            }
+
+            IdentityDatabaseUtil.closeAllConnections(null, null, prepStmt);
```

Comment:
closing prepared statement should be done in finally block. as it is done already, seems to be there is no need to do it here